File: app/src/main/java/ru/nikky/notes/ui/pages/edit/EditNoteFragment.java
package ru.nikky.notes.ui.pages.edit;

import android.content.Context;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.EditText;

import com.google.android.material.button.MaterialButton;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import ru.nikky.notes.R;
import ru.nikky.notes.databinding.FragmentEditNoteBinding;
import ru.nikky.notes.domain.NoteEntity;

public class EditNoteFragment extends Fragment {

    private FragmentEditNoteBinding binding;
    public static final String KEY_NOTE_ENTITY = "KEY_NOTE_ENTITY";
    private boolean isNewNote;
    private NoteEntity inputNoteEntity;

    @Override
    public void onAttach(@NonNull Context context) {
        super.onAttach(context);
        if (!(context instanceof EditNoteFragment.Contract)){
            throw new IllegalStateException("Launcher activity must implement EditNoteFragment.Contract");
        }
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        binding = FragmentEditNoteBinding.inflate(inflater, container, false);
        return binding.getRoot();
    }

    @Override
    public void onDestroyView() {
        binding = null;
        super.onDestroyView();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        setupListeners();
        processInputArguments(getArguments());
    }

    private void setupListeners() {
        binding.saveNoteButton.setOnClickListener(v -> getContractActivity().saveResult(getNote()));
    }

    private NoteEntity getNote() {
        String titleText = binding.titleEditText.getText().toString().trim();
        String detailText = binding.detailEditText.getText().toString().trim();
        NoteEntity outputNoteEntity;
        if (isNewNote){
            outputNoteEntity = new NoteEntity(titleText, detailText);
        } else {
            inputNoteEntity.setTitle(titleText);
            inputNoteEntity.setDetail(detailText);
            outputNoteEntity = inputNoteEntity;
        }
        return outputNoteEntity;
    }

    private void processInputArguments(Bundle arguments) {
        inputNoteEntity = (NoteEntity) arguments.get(KEY_NOTE_ENTITY);
        isNewNote = inputNoteEntity == null;
        if (!isNewNote){
            binding.titleEditText.setText(inputNoteEntity.getTitle());
            binding.detailEditText.setText(inputNoteEntity.getDetail());
        }
    }

    public static EditNoteFragment newInstance(NoteEntity noteEntity){
        EditNoteFragment editNoteFragment = new EditNoteFragment();
        Bundle bundle = new Bundle();
        bundle.putParcelable(KEY_NOTE_ENTITY, noteEntity);
        editNoteFragment.setArguments(bundle);
        return editNoteFragment;
    }

    public interface Contract{
        void saveResult(NoteEntity noteEntity);
    }

    private Contract getContractActivity(){
        return (EditNoteFragment.Contract) getActivity();
    }
}

File: app/src/main/java/ru/nikky/notes/ui/pages/edit/EditNoteFragment.kt
package ru.nikky.notes.ui.pages.edit

import android.content.Context
import android.os.Bundle
import android.view.View
import androidx.fragment.app.Fragment
import by.kirich1409.viewbindingdelegate.viewBinding
import ru.nikky.notes.R
import ru.nikky.notes.databinding.FragmentEditNoteBinding
import ru.nikky.notes.domain.NoteEntity

class EditNoteFragment : Fragment(R.layout.fragment_edit_note) {
    private val binding: FragmentEditNoteBinding by viewBinding(FragmentEditNoteBinding::bind)
    private var isNewNote = false
    private var inputNoteEntity: NoteEntity? = null
    val currentNoteId: Int
        get() = if (inputNoteEntity == null) -1 else inputNoteEntity!!.id

    override fun onAttach(context: Context) {
        super.onAttach(context)
        check(context is Contract) { "Launcher activity must implement EditNoteFragment.Contract" }
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        setupListeners()
        processInputArguments()
    }

    private fun setupListeners() {
        binding.saveNoteButton.setOnClickListener { contractActivity.saveResult(note) }
    }

    private val note: NoteEntity?
        get() {
            val titleText = binding.titleEditText.text.toString().trim { it <= ' ' }
            val detailText = binding.detailEditText.text.toString().trim { it <= ' ' }
            val outputNoteEntity = if (isNewNote) {
                NoteEntity(titleText, detailText)
            } else {
                inputNoteEntity!!.title = titleText
                inputNoteEntity!!.detail = detailText
                inputNoteEntity
            }
            return outputNoteEntity
        }

    private fun processInputArguments() {
        inputNoteEntity = requireArguments()[KEY_NOTE_ENTITY] as NoteEntity?
        isNewNote = inputNoteEntity == null
        if (!isNewNote) {
            binding.titleEditText.setText(inputNoteEntity!!.title)
            binding.detailEditText.setText(inputNoteEntity!!.detail)
        }
    }

    interface Contract {
        fun saveResult(noteEntity: NoteEntity?)
    }

    private val contractActivity: Contract
        get() = activity as Contract

    companion object {
        const val KEY_NOTE_ENTITY = "KEY_NOTE_ENTITY"

        @JvmStatic
        fun newInstance(noteEntity: NoteEntity?): EditNoteFragment {
            val editNoteFragment = EditNoteFragment()
            val bundle = Bundle()
            bundle.putParcelable(KEY_NOTE_ENTITY, noteEntity)
            editNoteFragment.arguments = bundle
            return editNoteFragment
        }
    }
}